db/db.go

```diff
@@ -16,52 +25,225 @@ type Scannable interface {
+	return newDealAccessor(d.db, deal)
+}
+
+func newDealAccessor(db *sql.DB, deal *types.ProviderDealState) *dealAccessor {
+	return &dealAccessor{
+		db:   db,
+		deal: deal,
+		def: map[string]fieldDefinition{
+			"ID":                    &fieldDef{f: &deal.DealUuid},
+			"CreatedAt":             &fieldDef{f: &deal.CreatedAt},
+			"DealProposalSignature": &sigFieldDef{f: &deal.ClientDealProposal.ClientSignature},
+			"PieceCID":              &cidFieldDef{f: &deal.ClientDealProposal.Proposal.PieceCID},
+			"PieceSize":             &fieldDef{f: &deal.ClientDealProposal.Proposal.PieceSize},
+			"VerifiedDeal":          &fieldDef{f: &deal.ClientDealProposal.Proposal.VerifiedDeal},
+			"ClientAddress":         &addrFieldDef{f: &deal.ClientDealProposal.Proposal.Client},
+			"ProviderAddress":       &addrFieldDef{f: &deal.ClientDealProposal.Proposal.Provider},
+			"Label":                 &fieldDef{f: &deal.ClientDealProposal.Proposal.Label},
+			"StartEpoch":            &fieldDef{f: &deal.ClientDealProposal.Proposal.StartEpoch},
+			"EndEpoch":              &fieldDef{f: &deal.ClientDealProposal.Proposal.EndEpoch},
+			"StoragePricePerEpoch":  &bigIntFieldDef{f: &deal.ClientDealProposal.Proposal.StoragePricePerEpoch},
+			"ProviderCollateral":    &bigIntFieldDef{f: &deal.ClientDealProposal.Proposal.ProviderCollateral},
+			"ClientCollateral":      &bigIntFieldDef{f: &deal.ClientDealProposal.Proposal.ClientCollateral},
+			"SelfPeerID":            &fieldDef{f: &deal.SelfPeerID},
+			"ClientPeerID":          &fieldDef{f: &deal.ClientPeerID},
+			"DealDataRoot":          &cidFieldDef{f: &deal.DealDataRoot},
+			"InboundFilePath":       &fieldDef{f: &deal.InboundFilePath},
+			"TransferType":          &fieldDef{f: &deal.TransferType},
+			"TransferParams":        &fieldDef{f: &deal.TransferParams},
+			"ChainDealID":           &fieldDef{f: &deal.ChainDealID},
+			"PublishCID":            &cidFieldDef{f: deal.PublishCID},
+			"SectorID":              &fieldDef{f: &deal.SectorID},
+			"Offset":                &fieldDef{f: &deal.Offset},
+			"Length":                &fieldDef{f: &deal.Length},
+			"Checkpoint":            &ckptFieldDef{f: &deal.Checkpoint},
+			"Error":                 &fieldDef{f: &deal.Err},
+		},
+	}
+}
+
+func (d *dealAccessor) scan(row Scannable) error {
+	// For each field
+	dest := []interface{}{}
+	for _, name := range dealFields {
+		// Get a pointer to the field that will receive the scanned value
+		fieldDef := d.def[name]
+		dest = append(dest, fieldDef.fieldPtr())
+	}
+
+	// Scan the row into each pointer
+	err := row.Scan(dest...)
+	if err != nil {
+		return xerrors.Errorf("scanning deal row: %w", err)
+	}
+
+	// For each field
+	for _, fieldDef := range d.def {
+		// Unmarshall the scanned value into deal object
+		err := fieldDef.unmarshall()
+		if err != nil {
+			return xerrors.Errorf("unmarshalling db field: %s", err)
+		}
+	}
+	return nil
+}
+
+func (d *dealAccessor) insert(ctx context.Context) error {
```

Comment:
Does it error out if primary key already exists  ? Does this error out if the primary key is missing or does it generate it's own ?
 
 Given that the deal UUID is going to be generated by the client, is there a way to tell the DB to never generate it's own primary key and always mandate that the caller provides one ?

Reply:
It will error out if the primary key already exists because in the CREATE TABLE statement we have the line
```
PRIMARY KEY(ID)
```

> is there a way to tell the DB to never generate it's own primary key and always mandate that the caller provides one

The `PRIMARY KEY(ID)` will also take care of this